src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt

```diff
@@ -462,3 +409,64 @@ private fun resolveThroughTypeAliases(boundElement: BoundElement<RsElement>): Bo
     }
     return base
 }
+
+/**
+ * Consider this code:
+ *
+ * ```rust
+ * trait Trait {
+ *     type Type;
+ * }
+ * struct S;
+ * impl Trait for S {
+ *     type Type = ();
+ * }
+ * type A = <S as Trait>::Type;
+ * ```
+ *
+ * Usual `path.reference.resolve()` always resolves `<S as Trait>::Type` path
+ * to the associated type in the trait (`trait Trait { type Type; }`). Such behavior is handy
+ * for type inference, but unhandy for users (that most likely want to go to declaration in the `impl`)
+ * and for other clients of name resolution (like intention actions).
+ *
+ * Here we're trying to find concrete `impl` for resolved associated type.
+ */
+private fun RsPathReference.tryAdvancedResolveTypeAliasToImpl(): BoundElement<RsElement>? {
+    val resolvedBoundElement = advancedResolve() ?: return null
+
+    // 1. Check that we're resolved to an associated type inside a trait:
+    // trait Trait {
+    //     type Item;
+    // }      //^ resolved here
+    val resolved = resolvedBoundElement.element as? RsTypeAlias ?: return null
+    if (resolved.owner !is RsAbstractableOwner.Trait) return null
+
+    // 2. Check that we resolve a `Self`-qualified path or explicit type-qualified path:
+    //    `Self::Type` or `<Foo as Trait>::Type`
+    val typeQual = element.typeQual
+    if (element.path?.hasCself != true && (typeQual == null || typeQual.traitRef == null)) return null
+
+    // 3. Try to select a concrete impl for the associated type
+    val lookup = ImplLookup.relativeTo(element)
+    val selection = lookup.selectStrict(
+        (TyProjection.valueOf(resolved).substitute(resolvedBoundElement.subst) as TyProjection).traitRef
+    ).ok()
+    if (selection?.impl !is RsImplItem) return null
+    val element = selection.impl.expandedMembers.types.find { it.name == resolved.name } ?: return null
+    val newSubst = lookup.ctx.fullyResolveWithOrigins(selection.subst)
+    NameResolutionTestmarks.TypeAliasToImpl.hit()
+    return BoundElement(element, newSubst)
+}
+
+/** @see tryAdvancedResolveTypeAliasToImpl */
+fun RsPathReference.advancedResolveTypeAliasToImpl(): BoundElement<RsElement>? =
+    tryAdvancedResolveTypeAliasToImpl() ?: advancedResolve()
+
+/** @see tryAdvancedResolveTypeAliasToImpl */
+fun RsPathReference.tryResolveTypeAliasToImpl(): RsElement? =
+    tryAdvancedResolveTypeAliasToImpl()?.element
+
+/** @see tryAdvancedResolveTypeAliasToImpl */
+fun RsPathReference.resolveTypeAliasToImpl(): RsElement? {
+    return tryResolveTypeAliasToImpl() ?: resolve()
```

Comment:
Is it ok that we will call `resolve` twice if `tryResolveTypeAliasToImpl` is null?

Reply:
Since the value is cached, I think it's ok. Sometimes even cache access can be a bottleneck, but this method is used in documentation link name resolution code, so it can't be a bottleneck at all.